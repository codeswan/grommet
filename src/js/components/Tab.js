// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React, { Component, PropTypes } from 'react';
import KeyboardAccelerators from '../utils/KeyboardAccelerators';
import CSSClassnames from '../utils/CSSClassnames';

const CLASS_ROOT = CSSClassnames.TAB;

export default class Tab extends Component {

  constructor(props, context) {
    super(props, context);

    this._processSpace = this._processSpace.bind(this);
    this._onClickTab = this._onClickTab.bind(this);
  }

  componentDidMount () {
    KeyboardAccelerators.startListeningToKeyboard(this, {
      space: this._processSpace
    });
  }

  componentWillUnmount () {
    KeyboardAccelerators.stopListeningToKeyboard(this, {
      space: this._processSpace
    });
  }

  _processSpace (event) {
    if (event.target === this.tabRef) {
      this._onClickTab(event);
    }
  }

  _onClickTab (event) {
    if (event) {
      event.preventDefault();
    }
    this.props.onRequestForActive();
  }

  render () {
    var classes = [CLASS_ROOT];

    if (this.props.active) {
      classes.push(CLASS_ROOT + "--active");
    }

    return (
      <li className={classes.join(' ')} id={this.props.id}>
        <a ref={(ref) => this.tabRef = ref} role="tab"
          href="#" onClick={this._onClickTab}
          aria-expanded={this.props.active} aria-selected={this.props.active}
          className={CLASS_ROOT + "__link"} aria-labelledby={this.props.id}>
          <label className={CLASS_ROOT + '__label'} htmlFor={this.props.id}>
            {this.props.title}
          </label>
        </a>
      </li>
    );
  }

}

Tab.propTypes = {
  title: PropTypes.string.isRequired,
  active: PropTypes.bool,
  id: PropTypes.string
};
