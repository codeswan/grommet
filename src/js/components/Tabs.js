// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import Intl from '../utils/Intl';
import Box from './Box';
import CSSClassnames from '../utils/CSSClassnames';

const CLASS_ROOT = CSSClassnames.TABS;

export default class Tabs extends Component {

  constructor(props, context) {
    super(props, context);

    this._activateTab = this._activateTab.bind(this);

    this.state = {
      activeIndex: props.initialIndex,
      justify: props.justify
    };
  }

  componentWillReceiveProps(nextProps) {
    if ((nextProps.activeIndex || 0 === nextProps.activeIndex) &&
      this.state.activeIndex !== nextProps.activeIndex) {
      this.setState({activeIndex: nextProps.activeIndex});
    }
  }

  _activateTab (index) {
    this.setState({activeIndex: index});
  }

  render () {
    let classes = classnames(
      CLASS_ROOT,
      this.props.className,
      {
        [`${CLASS_ROOT}--justify-${this.props.justify}`]: this.props.justify,
        [`${CLASS_ROOT}--responsive`]: this.props.responsive
      }
    );

    var activeContainer;
    var activeTitle;

    var tabs = React.Children.map(this.props.children, function(tab, index) {

      var tabProps = tab.props || tab._store.props || {};

      var isTabActive = index === this.state.activeIndex;

      if (isTabActive) {
        activeContainer = tabProps.children;
        activeTitle = tabProps.title;
      }

      return React.cloneElement(tab, {
        active: isTabActive,
        id: 'tab-' + index,
        onRequestForActive: function () {
          this._activateTab(index);
        }.bind(this)
      });
    }.bind(this));

    var tabContentTitle = Intl.getMessage(this.context.intl, 'Tab Contents', {
      activeTitle: activeTitle
    });

    //TODO: Since there could be multiple Tabs on the page, we need a more
    //robust means of identifying the association between title and aria label.
    return (
      <div role="tablist">
        <ul className={classes}>
          {tabs}
        </ul>
        <div tabIndex="0"
          aria-label={tabContentTitle} role="tabpanel">
          <Box className={CLASS_ROOT + '__content'}
            aria-label={tabContentTitle}>
            {activeContainer}
          </Box>
        </div>
      </div>
    );
  }
}

Tabs.propTypes = {
  activeIndex: PropTypes.number,
  justify: PropTypes.oneOf(['start', 'center', 'end']),
  responsive: PropTypes.bool
};

Tabs.contextTypes = {
  intl: PropTypes.object
};

Tabs.defaultProps = {
  initialIndex: 0,
  justify: 'center',
  responsive: true
};
