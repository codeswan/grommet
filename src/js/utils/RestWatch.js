// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import Rest from './Rest';

const RECONNECT_TIMEOUT = 5000; // 5s
const POLL_TIMEOUT = 10000; // 10s

let state = {
  ws: null,
  wsReady: false,
  timer: null,
  requests: [], // {message: , context: }
  nextRequestId: 1,
  initialized: false,
  socketUrl: null
};

export default {

  _sendMessage (op, id, url, params) {
    state.ws.send(JSON.stringify({
      op: op,
      id: id,
      url: url,
      params: params
    }));
  },

  _onOpen () {
    state.wsReady = true;
    // send any requests we have queued up
    state.requests.forEach(function(request) {
      this._sendMessage('start', request.id, request.url, request.params);
    }, this);
  },

  _onError (error) {
    console.log('!!! RestWatch _onError TODO', error);
  },

  _onMessage (event) {
    var message = JSON.parse(event.data);
    // Figure out which message this corresponds to so we
    // know which action to deliver the response with.
    state.requests.some(function(request) {
      if (request.id === message.id) {
        request.handler(message.result);
      }
    });
  },

  _onClose () {
    // lost connection, retry in a bit
    state.ws = null;
    state.wsReady = false;
    state.initialized = false;
    state.timer = setTimeout(this.initialize.bind(this), RECONNECT_TIMEOUT);
  },

  _getREST (request) {
    request.pollBusy = true;
    Rest.get(request.url, request.params)
      .end(function(err, res) {
        if (err) {
          throw err;
        }

        if (res.ok) {
          request.handler(res.body);
        }
        request.pollBusy = false;
      });
  },

  _poll () {
    state.requests.forEach(function(request) {
      if (!request.pollBusy) {
        this._getREST(request);
      }
    });
  },

  initialize (socketUrl) {
    if (!state.initialized && !state.ws && this.available() &&
      (state.socketUrl || socketUrl)) {
      state.socketUrl = state.socketUrl || socketUrl;
      state.ws = new WebSocket(state.socketUrl);
      state.ws.onopen = this._onOpen.bind(this);
      state.ws.onerror = this._onError.bind(this);
      state.ws.onmessage = this._onMessage.bind(this);
      state.ws.onclose = this._onClose.bind(this);
      state.initialized = true;
    }
  },

  available () {
    return ('WebSocket' in window || 'MozWebSocket' in window);
  },

  start (url, params, handler) {
    this.initialize();
    var request = {
      id: state.nextRequestId,
      url: url,
      params: params,
      handler: handler
    };
    state.nextRequestId += 1;
    state.requests.push(request);

    if (state.wsReady) {
      this._sendMessage('start', request.id, request.url, request.params);
    } else if (!this.available()) {
      // no web sockets, fall back to polling
      this._getREST(request);
      clearTimeout(state.timer);
      state.timer = setTimeout(this._poll.bind(this), POLL_TIMEOUT);
    }
    return request.id;
  },

  stop (requestId) {
    state.requests = state.requests.filter(function(request) {
      if (request.id === requestId) {
        if (state.wsReady) {
          this._sendMessage('stop', request.id);
        }
        return false;
      } else {
        return true;
      }
    }, this);
  }
};
