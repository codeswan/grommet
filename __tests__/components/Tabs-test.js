// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import Tabs from '../../src/js/components/Tabs';
import Tab from '../../src/js/components/Tab';

import { findAllByType } from '../utils/renderer-finder';

// needed because this:
// https://github.com/facebook/jest/issues/1353
jest.mock('react-dom');

describe('Tabs', () => {
  it('has correct default options', () => {
    const component = renderer.create(
      <Tabs>
        <Tab title="Tab Title">
          <h1>test</h1>
        </Tab>
        <Tab title="Tab Title 2">
          <h1>test 2</h1>
        </Tab>
      </Tabs>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('has correct initialIndex=0 rendering', () => {
    const component = renderer.create(
      <Tabs activeIndex={0}>
        <Tab title="Tab Title">
          <h1>test</h1>
        </Tab>
        <Tab title="Tab Title 2">
          <h1>test 2</h1>
        </Tab>
      </Tabs>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // activating the second panel
    const panelHeaders = findAllByType(tree, 'a');
    panelHeaders[1].props.onClick();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
