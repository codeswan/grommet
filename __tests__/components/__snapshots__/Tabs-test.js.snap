exports[`Tabs has correct default options 1`] = `
<div
  role="tablist">
  <ul
    className="grommetux-tabs grommetux-tabs--justify-center grommetux-tabs--responsive">
    <li
      className="grommetux-tab grommetux-tab--active"
      id="tab-0">
      <a
        aria-expanded={true}
        aria-labelledby="tab-0"
        aria-selected={true}
        className="grommetux-tab__link"
        href="#"
        onClick={[Function bound _onClickTab]}
        role="tab">
        <label
          className="grommetux-tab__label"
          htmlFor="tab-0">
          Tab Title
        </label>
      </a>
    </li>
    <li
      className="grommetux-tab"
      id="tab-1">
      <a
        aria-expanded={false}
        aria-labelledby="tab-1"
        aria-selected={false}
        className="grommetux-tab__link"
        href="#"
        onClick={[Function bound _onClickTab]}
        role="tab">
        <label
          className="grommetux-tab__label"
          htmlFor="tab-1">
          Tab Title 2
        </label>
      </a>
    </li>
  </ul>
  <div
    aria-label="Tab Contents"
    role="tabpanel"
    tabIndex="0">
    <div
      aria-label="Tab Contents"
      className="grommetux-box grommetux-box--direction-column grommetux-box--responsive grommetux-box--pad-none grommetux-tabs__content"
      id={undefined}
      onClick={undefined}
      role={undefined}
      style={Object {}}
      tabIndex={undefined}>
      <h1>
        test
      </h1>
    </div>
  </div>
</div>
`;

exports[`Tabs has correct initialIndex=0 rendering 1`] = `
<div
  role="tablist">
  <ul
    className="grommetux-tabs grommetux-tabs--justify-center grommetux-tabs--responsive">
    <li
      className="grommetux-tab grommetux-tab--active"
      id="tab-0">
      <a
        aria-expanded={true}
        aria-labelledby="tab-0"
        aria-selected={true}
        className="grommetux-tab__link"
        href="#"
        onClick={[Function bound _onClickTab]}
        role="tab">
        <label
          className="grommetux-tab__label"
          htmlFor="tab-0">
          Tab Title
        </label>
      </a>
    </li>
    <li
      className="grommetux-tab"
      id="tab-1">
      <a
        aria-expanded={false}
        aria-labelledby="tab-1"
        aria-selected={false}
        className="grommetux-tab__link"
        href="#"
        onClick={[Function bound _onClickTab]}
        role="tab">
        <label
          className="grommetux-tab__label"
          htmlFor="tab-1">
          Tab Title 2
        </label>
      </a>
    </li>
  </ul>
  <div
    aria-label="Tab Contents"
    role="tabpanel"
    tabIndex="0">
    <div
      aria-label="Tab Contents"
      className="grommetux-box grommetux-box--direction-column grommetux-box--responsive grommetux-box--pad-none grommetux-tabs__content"
      id={undefined}
      onClick={undefined}
      role={undefined}
      style={Object {}}
      tabIndex={undefined}>
      <h1>
        test
      </h1>
    </div>
  </div>
</div>
`;

exports[`Tabs has correct initialIndex=0 rendering 2`] = `
<div
  role="tablist">
  <ul
    className="grommetux-tabs grommetux-tabs--justify-center grommetux-tabs--responsive">
    <li
      className="grommetux-tab"
      id="tab-0">
      <a
        aria-expanded={false}
        aria-labelledby="tab-0"
        aria-selected={false}
        className="grommetux-tab__link"
        href="#"
        onClick={[Function bound _onClickTab]}
        role="tab">
        <label
          className="grommetux-tab__label"
          htmlFor="tab-0">
          Tab Title
        </label>
      </a>
    </li>
    <li
      className="grommetux-tab grommetux-tab--active"
      id="tab-1">
      <a
        aria-expanded={true}
        aria-labelledby="tab-1"
        aria-selected={true}
        className="grommetux-tab__link"
        href="#"
        onClick={[Function bound _onClickTab]}
        role="tab">
        <label
          className="grommetux-tab__label"
          htmlFor="tab-1">
          Tab Title 2
        </label>
      </a>
    </li>
  </ul>
  <div
    aria-label="Tab Contents"
    role="tabpanel"
    tabIndex="0">
    <div
      aria-label="Tab Contents"
      className="grommetux-box grommetux-box--direction-column grommetux-box--responsive grommetux-box--pad-none grommetux-tabs__content"
      id={undefined}
      onClick={undefined}
      role={undefined}
      style={Object {}}
      tabIndex={undefined}>
      <h1>
        test 2
      </h1>
    </div>
  </div>
</div>
`;
